glusterfsd/src/glusterfsd-mgmt.c

```diff
@@ -2330,6 +2330,13 @@ mgmt_getspec_cbk(struct rpc_req *req, struct iovec *iov, int count,
         cleanup_and_exit(0);
     }
 
+    if (ENOENT == ret) {
```

Comment:
This change is not required , if ret value is negative the condition exists at line 2340 is sufficient to exit a client process.

Reply:
Thanks a lot. But if I only set op_errno in __server_getspec. I find the glusterfs parent process blocks in sys_read(ctx->daemon_pipe[0]), and the  child process blocks in gf_fuse_umount during the exit program. So I add emancipate(ctx, ret) into glusterfsd-mgmt.c.

Reply:
when I only set op_errno in __server_getspec and run mount -t glusterfs  192.168.123.68:gv /root/zy, the client will not exit successfully and the log of client is as follows:
[2021-02-05 07:17:12.858644] I [MSGID: 100030] [glusterfsd.c:2689:main] 0-/usr/sbin/glusterfs: Started running version [{arg=/usr/sbin/glusterfs}, {version=8.2}, {cmdlinestr=/usr/sbin/glusterfs --process-name fuse --volfile-server=192.168.123.68 --volfile-id=gv /root/zy}]
[2021-02-05 07:17:12.859934] I [glusterfsd.c:2424:daemonize] 0-glusterfs: Pid of current running process is 3716
[2021-02-05 07:17:12.871348] I [MSGID: 101190] [event-epoll.c:670:event_dispatch_epoll_worker] 0-epoll: Started thread with index [{index=0}]
[2021-02-05 07:17:12.871459] I [MSGID: 101190] [event-epoll.c:670:event_dispatch_epoll_worker] 0-epoll: Started thread with index [{index=1}]
[2021-02-05 07:17:12.872517] E [glusterfsd-mgmt.c:2137:mgmt_getspec_cbk] 0-glusterfs: failed to get the 'volume file' from server
[2021-02-05 07:17:12.872591] E [glusterfsd-mgmt.c:2338:mgmt_getspec_cbk] 0-mgmt: failed to fetch volume file (key:gv)
[2021-02-05 07:17:12.873150] W [glusterfsd.c:1439:cleanup_and_exit] (-->/lib64/libgfrpc.so.0(+0xf101) [0x7f2ab0efc101] -->/usr/sbin/glusterfs(mgmt_getspec_cbk+0x961) [0x55929ca55ac1] -->/usr/sbin/glusterfs(cleanup_and_exit+0x6b) [0x55929ca4d9ab] ) 0-: received signum (0), shutting down
[2021-02-05 07:17:12.873222] I [fuse-bridge.c:7074:fini] 0-fuse: Unmounting '/root/zy'.
[2021-02-05 07:17:15.871277] I [timer.c:86:gf_timer_call_cancel] (-->/lib64/libglusterfs.so.0(+0x374a8) [0x7f2ab11604a8] -->/lib64/libgfrpc.so.0(rpc_clnt_reconnect+0x3b) [0x7f2ab0efb6db] -->/lib64/libglusterfs.so.0(gf_timer_call_cancel+0x149) [0x7f2ab11609e9] ) 0-timer: ctx cleanup started
[2021-02-05 07:17:22.871774] I [timer.c:86:gf_timer_call_cancel] (-->/lib64/libglusterfs.so.0(+0x374a8) [0x7f2ab11604a8] -->/lib64/libgfrpc.so.0(+0xfe47) [0x7f2ab0efce47] -->/lib64/libglusterfs.so.0(gf_timer_call_cancel+0x149) [0x7f2ab11609e9] ) 0-timer: ctx cleanup started
[2021-02-05 07:17:22.871847] E [timer.c:34:gf_timer_call_after] (-->/lib64/libglusterfs.so.0(+0x374a8) [0x7f2ab11604a8] -->/lib64/libgfrpc.so.0(+0xfe64) [0x7f2ab0efce64] -->/lib64/libglusterfs.so.0(gf_timer_call_after+0x229) [0x7f2ab1160729] ) 0-timer: Either ctx is NULL or ctx cleanup started [Invalid argument]
[2021-02-05 07:17:22.871877] W [rpc-clnt.c:148:call_bail] 0-glusterfs: Cannot create bailout timer for 192.168.123.68:24007

Reply:
You can call emancipate before cleanup_and_exit(0) is calling at line 2346.

Reply:
Thanks for your advice. Do you have the same problem in your version 8.2？

Reply:
No, i tried to reproduce on my laptop(fedora-32 glusterfs-8.2) i was not able to reproduce it.

Reply:
Thanks for your help. mine is centos-32 glusterfs-8.2.

Reply:
Is it possible for you to upload a test case?

Reply:
I am not familiar with it. Can you tell me how to do?

Reply:
You can follow the document glusterfs/tests/README how to run a test case.You can run test case like 
prove -vf tests/basic/volume.t for any test case after install the glusterfs bits in your environment.